### PR TITLE
Update status and code index after changing the default branch

### LINKED ISSUE #26723: conditions that code cannot be searched
### Description

**BUG**

cannot search code

- or when code is within non-newest commit

- or when code is within non-default branch

- or when code is within non-default branch X, but now X is default branch

**reproduce**

- copy paste indexer config from [docs](https://docs.gitea.com/administration/repo-indexer)

- `echo '"HELLO"' > reproduce.js`, commit

- `echo '"WORLD"' > reproduce.js`, commit

- push to **non-default** branch `test`

- stop gitea, rm `indexers/repos.bleve`, start gitea

- ensure default branch is `main`

- search `HELLO`, got nothing, **which is unexpectedly**

- search `WORLD`, got nothing, **which is unexpectedly**

- update default branch from `main` to `test`

- search `HELLO`, got nothing, **which is unexpectedly**

- search `WORLD`, got nothing, **which is unexpectedly**

- stop gitea, rm `indexers/repos.bleve`, start gitea

- search `HELLO`, got nothing, **which is unexpectedly**

- search `WORLD`, got match

- update default branch from `test` to `main`

- stop gitea, rm `indexers/repos.bleve`, start gitea

### Gitea Version

1.20.3

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

docker image `docker.io/gitea/gitea:1.20.3`

### Database

None

### PR DESCRIPTION
**Pull request by nanguanlin6**
Fix #26723 
Add `ChangeDefaultBranch`  to the `notifier` interface and implement it in `indexerNotifier`. So when changing the default branch, `indexerNotifier` sends a message to the `indexer queue` to update the index.

### CODE REVIEW

File: modules/indexer/code/git.go [hunk 1]
```diff
@@ -30,7 +30,18 @@ func getRepoChanges(ctx context.Context, repo *repo_model.Repository, revision s
 		return nil, err
 	}
 
+	var needGenesis bool
 	if len(status.CommitSha) == 0 {
+		needGenesis = true
+	}
+
```
Comment (lunny) [inline modules/indexer/code/git.go]:
```suggestion
if !needGenesis {
```

File: routers/web/repo/setting/default_branch.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
+package setting
```
Comment (lunny) [inline routers/web/repo/setting/default_branch.go:4]: copyright header

Comment (techknowlogick) [reply routers/web/repo/setting/default_branch.go:4]:
```suggestion

package setting
```

File: modules/indexer/code/git.go [hunk 3]
```diff
@@ -30,7 +30,20 @@ func getRepoChanges(ctx context.Context, repo *repo_model.Repository, revision s
...
+	if !needGenesis {
+		hasAncestorCmd := git.NewCommand(ctx, "merge-base").AddDynamicArguments(repo.CodeIndexerStatus.CommitSha, revision)
+		stdout, _, _ := hasAncestorCmd.RunStdString(&git.RunOpts{Dir: repo.RepoPath()})
+		if stdout == "" {
+			needGenesis = true
+		}
```
Comment (lunny) [inline modules/indexer/code/git.go]:
```suggestion
    needGenesis = stdout == ""
```

**Review by lunny — APPROVED** [review]

**Review by wxiaoguang — APPROVED** [review]

**Pull request merged**